File: projects/advent_of_code/solutions/year2015/day03/python/solution.py
#!/usr/bin/env python3
"""Advent of Code 2015 Day 3"""

from projects.advent_of_code.python.aoc import read_file, Point, UP, DOWN, LEFT, RIGHT


def part1(input_text: str) -> int:
    directions = {
        "^": UP,
        "v": DOWN,
        "<": LEFT,
        ">": RIGHT,
    }

    visited = {Point(0, 0)}
    pos = Point(0, 0)

    for char in input_text.strip():
        pos = pos + directions[char]
        visited.add(pos)

    return len(visited)


def part2(input_text: str) -> int:
    directions = {
        "^": UP,
        "v": DOWN,
        "<": LEFT,
        ">": RIGHT,
    }

    visited = {Point(0, 0)}
    santa_pos = Point(0, 0)
    robo_pos = Point(0, 0)

    for i, char in enumerate(input_text.strip()):
        if i % 2 == 0:
            santa_pos = santa_pos + directions[char]
            visited.add(santa_pos)
        else:
            robo_pos = robo_pos + directions[char]
            visited.add(robo_pos)

    return len(visited)


if __name__ == "__main__":
    input_text = read_file("../input.txt")

    print(f"Part 1: {part1(input_text)}")
    print(f"Part 2: {part2(input_text)}")

File: projects/advent_of_code/solutions/year2015/day03/go/solution.go
package main

import (
	"fmt"
	"strings"

	"github.com/jomcgi/homelab/projects/advent_of_code/pkg/aoc"
)

func main() {
	input, err := aoc.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}

	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}

func part1(input string) int {
	directions := map[rune]aoc.Point{
		'^': aoc.Up,
		'v': aoc.Down,
		'<': aoc.Left,
		'>': aoc.Right,
	}

	visited := map[aoc.Point]bool{{X: 0, Y: 0}: true}
	pos := aoc.Point{X: 0, Y: 0}

	for _, char := range strings.TrimSpace(input) {
		pos = pos.Add(directions[char])
		visited[pos] = true
	}

	return len(visited)
}

func part2(input string) int {
	directions := map[rune]aoc.Point{
		'^': aoc.Up,
		'v': aoc.Down,
		'<': aoc.Left,
		'>': aoc.Right,
	}

	visited := map[aoc.Point]bool{{X: 0, Y: 0}: true}
	santaPos := aoc.Point{X: 0, Y: 0}
	roboPos := aoc.Point{X: 0, Y: 0}

	for i, char := range strings.TrimSpace(input) {
		if i%2 == 0 {
			santaPos = santaPos.Add(directions[char])
			visited[santaPos] = true
		} else {
			roboPos = roboPos.Add(directions[char])
			visited[roboPos] = true
		}
	}

	return len(visited)
}
